design/acme-orders-challenges-crd.md

```diff
@@ -0,0 +1,388 @@
+	DNSName  string `json:"dnsName"`
+	Token    string `json:"token"`
+	Key      string `json:"key"`
+	Wildcard bool   `json:"wildcard"`
+
+	// Config specifies the solver configuration for this challenge.
+	Config SolverConfig `json:"config"`
+
+	// IssuerRef references a properly configured ACME-type Issuer which should
+	// be used to create this Challenge.
+	// If the Issuer does not exist, processing will be retried.
+	// If the Issuer is not an 'ACME' Issuer, an error will be returned and the
+	// Challenge will be marked as failed.
+	IssuerRef ObjectReference `json:"issuerRef"`
+}
+
+type ChallengeStatus struct {
+	Presented bool   `json:"presented"`
+	Reason    string `json:"reason"`
+	State     State  `json:"state"`
+}
+```
+
+#### Changes to existing types
+
+The scope of changes to existing types as been purposely kept as small as possible.
+
+We do however require a change to the `status` field of the Certificate resource
+type. This should not be too problematic for users, as the cert-manager controller
+should be able to tolerate this change seamlessly.
+
+```go
+// CertificateStatus defines the observed state of Certificate
+type CertificateStatus struct {
+	Conditions      []CertificateCondition `json:"conditions,omitempty"`
+    ACME            *CertificateACMEStatus `json:"acme,omitempty"`
+    /////// NEW FIELD ///////
+	LastFailureTime *metav1.Time           `json:"lastFailureTime,omitempty"`
+}
+
+// CertificateACMEStatus holds the status for an ACME issuer
+type CertificateACMEStatus struct {
+    /////// REMOVED ALL OLD FIELDS ///////
+
+    /////// NEW FIELD ///////
+	// Order contains details about the current in-progress ACME Order.
+	// If this field is not set, an Order is not in progress.
+	// This field may point to a failed or inactive Order.
+	// It is not sufficient to check for the presence of this field in order to
+	// determine whether an order is in progress.
+	// +optional
+	OrderRef *LocalObjectReference `json:"orderRef,omitempty"`
+}
+```
+
+### Controllers
+
+Two new controllers will be built to reconcile the new `Order` and `Challenge`
+types.
+
+This two controllers will take the majority of the logic that used to exist in
+the `Prepare` function of the ACME issuer.
+
+#### Order controller
+
+When an order is created, all of the spec fields must be defined.
+The Order controller will be responsible for handling the entirety of an ACME
+order.
+The 'spec' describes an order to be created. The controller will keep the 'status'
+fields up to date by periodically syncing with the ACME server.
+
+* The Order controller will attempt to create a new order with the ACME server
+if the `status.url` field is not set. It will copy details of the order back
+onto the Order resoure, to save subsequent calls to the ACME server when not required.
+
+* If the orders 'state' is 'final' (i.e. one of valid, invalid or expired) then
+the controller will take no further action and return. (TODO: should we delete challenge
```

Comment:
Once the order is 'final', how long does it stay in the cluster? If it is indefinitely, then some users may never realize that they have these resources in their cluster

I think both of these resources should have a definable limit, so they can be cleaned up within a reasonable time frame(this could be user defined). And it would make sense that these controllers would handle that lifecycle

Reply:
👍 good observation - I have forgotten to mention that as part of the proposal!

The thinking is to have a 'spec.timeout' field - if the order or challenge has not completed within the timeout, then that resource will be updated/marked as timed out.

All Challenge resources will be retained for the lifetime if their owning *Order*.
All Order resources will be retained until a new order is required to be created (i.e. at renewal time).

We retain the Order resource so that we can get GetCertificate on the order if the user deletes the TLS certificate. This will save us having to create a new order each time, and allows us to instead retrieve the existing certificate from the ACME server.

I'll update the proposal accordingly!

Reply:
FWIW, if the order is *invalid*, it will be deleted and a new one created (after applying order-retry rate-limiting/back-off)

Reply:
> Will Order and Challenge resources also get deleted when the certificate is?
Yes

> If there is a limit to 1 order and 1 challenge per certificate, then I am much less concerned.
There should only ever be one order for each ACME Certificate at any one time, yes.

There will be N Challenge resources created for each certificate - where N is the number of domains on the Certificate.

> `All Challenge resources will be retained for the lifetime of their owning Order.`
Could potentially change. We may be best off actually deleting the challenge resource once the parent order associated with it is in a final state.